### PR TITLE
fuzz: add target for `DescriptorScriptPubKeyMan`

### PR DESCRIPTION
**Pull request by brunoerg**
This PR adds fuzz target for `DescriptorScriptPubKeyMan`. Also, moves `MockedDescriptorConverter` to `fuzz/util/descriptor` to be used here and in `descriptor` target.

### CODE REVIEW

**Review by maflcko — APPROVED** [review]
concept ack

File: src/wallet/test/fuzz/scriptpubkeyman.cpp
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: fixed

File: src/wallet/test/fuzz/scriptpubkeyman.cpp
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp:26]: Good question, let me check. If they are similar, we can move `MockedDescriptorConverter` to a util file and use it here.

Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: done!

File: src/wallet/test/fuzz/scriptpubkeyman.cpp
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: done

**Comment by dergoegge** [comment]
```bash
$ echo "cGsoJTg0KVwzKVwAQWlwZTgzKVwAQQAAAAAAADmDg4ODg4ODg4ODgYODg4ODg4Nwa2hhe3t7g4ODg4ODg4OD" | base64 --decode > scriptpubkeyman.crash
$ FUZZ=scriptpubkeyman ./src/test/fuzz/fuzz scriptpubkeyman.crash
...
Running: scriptpubkeyman.crash
fuzz: wallet/test/fuzz/scriptpubkeyman.cpp:93: auto wallet::(anonymous namespace)::scriptpubkeyman_fuzz_target(FuzzBufferType)::(anonymous class)::operator()() const: Assertion `spk_manager->GetDescriptorString(descriptor, fuzzed_data_provider.ConsumeBool())' failed.
==1391778== ERROR: libFuzzer: deadly signal
...
```

**Comment by brunoerg** [comment]
Thanks, @dergoegge. Fixed it by removing the `assert`, we can't ensure it.

**Comment by dergoegge** [comment]
```bash
$ echo "cGsoJTQxLyopXDMAXABhZGRyKGJjMYMxMTAwMDIwMDAwMDBnJyw2ZGRyKGI1bikp" | base64 --decode > scriptpubkeyman2.crash
$ FUZZ=scriptpubkeyman ./src/test/fuzz/fuzz scriptpubkeyman2.crash
...
fuzz: wallet/scriptpubkeyman.cpp:2008: virtual util::Result<CTxDestination> wallet::DescriptorScriptPubKeyMan::GetNewDestination(const OutputType): Assertion `desc_addr_type' failed.
==1499877== ERROR: libFuzzer: deadly signal
...
```

**Comment by brunoerg** [comment]
> I would recommend that you run the targets that you write for a while before opening a PR. It could very well be that it finds actual bugs that we don't want to be public right away.

Sure, sorry for that, left it running but not in my main machine so didn't get some of them. Force-pushed with a fix for `GetNewDestination`.  We can't fuzz it with random outputs types because any inconsistence will throw a runtime error. Any other one will return a `util::Error`. Left it running for some hours before pushing and will continue running it overnight.

**Review by Ayush170-Future — APPROVED** [review]
ACK

Reviewed the whole code. Looks great to me!

**Comment by dergoegge** [comment]
Here's a coverage report: https://dergoegge.github.io/bitcoin-coverage/pr28578/fuzz.coverage/index.html (~1000h CPU time).

Is there a reason not to cover more of `DescriptorScriptPubKeyMan` in this PR? e.g. `DescriptorScriptPubKeyMan::SignTransaction` or `DescriptorScriptPubKeyMan::FillPSBT`.

**Comment by brunoerg** [comment]
> Is there a reason not to cover more of DescriptorScriptPubKeyMan in this PR? e.g. DescriptorScriptPubKeyMan::SignTransaction or DescriptorScriptPubKeyMan::FillPSBT.

Not exactly, perhaps one of them may be very slow, let me test more functions and then I can add them all here, np.

**Comment by brunoerg** [comment]
Force-pushed covering more functions. Left it running for a long time and didn't see any downside.

File: src/wallet/test/fuzz/scriptpubkeyman.cpp
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp:146]: Yes, I'm addressing it.

Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: done

**Comment by brunoerg** [comment]
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/28578#discussion_r1389308579

File: src/wallet/test/fuzz/scriptpubkeyman.cpp
Comment (dergoegge) [inline src/wallet/test/fuzz/scriptpubkeyman.cpp:35]: I think this runs into the same timeouts as the `mocked_descriptor_parse` target. We should probably wait for #28832 to be merged so `HasDeepDerivPath` can be used here as well.

Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp:35]: I agree. I will draft this to wait for #28832.

**Comment by brunoerg** [comment]
To avoid timeouts in this target, I will leave it as draft until #28815 (or other solution) gets merged.

**Review by maflcko — COMMENTED** [review]
lgtm ACK 7774df96942ba10aeb5c818b100aedc9cb7d4b4f

File: src/test/fuzz/util/descriptor.h
Comment (brunoerg) [reply src/test/fuzz/util/descriptor.h]: done

File: src/wallet/test/fuzz/scriptpubkeyman.cpp
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: nop! done.

File: src/wallet/test/fuzz/scriptpubkeyman.cpp
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: done

**Comment by brunoerg** [comment]
force-pushed addressing:
https://github.com/bitcoin/bitcoin/pull/28578#discussion_r1399346490
https://github.com/bitcoin/bitcoin/pull/28578#discussion_r1399350026
https://github.com/bitcoin/bitcoin/pull/28578#discussion_r1399351848

**Review by dergoegge — APPROVED** [review]
ACK 47e5c9994c087d1826ccc0d539e916845b5648fb

**Comment by fanquake** [comment]
Null-dereference READ · wallet::scriptpubkeyman_fuzz_target 
```bash
Command: /mnt/scratch0/clusterfuzz/bot/builds/clusterfuzz-builds-honggfuzz_bitcoin-core_83aef6625aaeafa301867de74608b320f3c923fe/revisions/scriptpubkeyman
	Time ran: 0.5496230125427246
	
	Accepting input from '[STDIN]'
	Usage for fuzzing: honggfuzz -P [flags] -- /mnt/scratch0/clusterfuzz/bot/builds/clusterfuzz-builds-honggfuzz_bitcoin-core_83aef6625aaeafa301867de74608b320f3c923fe/revisions/scriptpubkeyman
	AddressSanitizer:DEADLYSIGNAL
	=================================================================
	==14133==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x79a1f8670aeb bp 0x7ffd95472db0 sp 0x7ffd95472578 T0)
	==14133==The signal is caused by a READ memory access.
	==14133==Hint: address points to the zero page.
	SCARINESS: 10 (null-deref)
	    #0 0x79a1f8670aeb in memcpy /build/glibc-SzIz7B/glibc-2.31/sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:142
	    #1 0x589ab4429a18 in __asan_memcpy /src/llvm-project/compiler-rt/lib/asan/asan_interceptors_memintrinsics.cpp:22:3
	    #2 0x589ab44cf869 in operator() [bitcoin-core/src/wallet/test/fuzz/scriptpubkeyman.cpp:0](https://github.com/bitcoin/bitcoin/blob/b5a271334ca81a6adcb1c608d85c83621a9eae47/src/wallet/test/fuzz/scriptpubkeyman.cpp#L0)
	    #3 0x589ab44cf869 in CallOneOf<(lambda at wallet/test/fuzz/scriptpubkeyman.cpp:75:13), (lambda at wallet/test/fuzz/scriptpubkeyman.cpp:87:13), (lambda at wallet/test/fuzz/scriptpubkeyman.cpp:94:13), (lambda at wallet/test/fuzz/scriptpubkeyman.cpp:108:13), (lambda at wallet/test/fuzz/scriptpubkeyman.cpp:117:13), (lambda at wallet/test/fuzz/scriptpubkeyman.cpp:121:13), (lambda at wallet/test/fuzz/scriptpubkeyman.cpp:135:13), (lambda at wallet/test/fuzz/scriptpubkeyman.cpp:149:13)> [bitcoin-core/src/test/fuzz/util.h:43](https://github.com/bitcoin/bitcoin/blob/b5a271334ca81a6adcb1c608d85c83621a9eae47/src/test/fuzz/util.h#L43):27
	    #4 0x589ab44cf869 in wallet::(anonymous namespace)::scriptpubkeyman_fuzz_target(Span<unsigned char const>) [bitcoin-core/src/wallet/test/fuzz/scriptpubkeyman.cpp:73](https://github.com/bitcoin/bitcoin/blob/b5a271334ca81a6adcb1c608d85c83621a9eae47/src/wallet/test/fuzz/scriptpubkeyman.cpp#L73):9
	    #5 0x589ab446d2d1 in __invoke<void (*&)(Span<const unsigned char>), Span<const unsigned char> > /usr/local/include/c++/v1/type_traits:3592:23
	    #6 0x589ab446d2d1 in __call<void (*&)(Span<const unsigned char>), Span<const unsigned char> > /usr/local/include/c++/v1/__functional/invoke.h:61:9
	    #7 0x589ab446d2d1 in operator() /usr/local/include/c++/v1/__functional/function.h:181:16
	    #8 0x589ab446d2d1 in std::__1::__function::__func<void (*)(Span<unsigned char const>), std::__1::allocator<void (*)(Span<unsigned char const>)>, void (Span<unsigned char const>)>::operator()(Span<unsigned char const>&&) /usr/local/include/c++/v1/__functional/function.h:355:12
	    #9 0x589ab49ef0b8 in operator() /usr/local/include/c++/v1/__functional/function.h:508:16
	    #10 0x589ab49ef0b8 in operator() /usr/local/include/c++/v1/__functional/function.h:1185:12
	    #11 0x589ab49ef0b8 in LLVMFuzzerTestOneInput [bitcoin-core/src/test/fuzz/fuzz.cpp:178](https://github.com/bitcoin/bitcoin/blob/b5a271334ca81a6adcb1c608d85c83621a9eae47/src/test/fuzz/fuzz.cpp#L178):5
	    #12 0x589ab642bdeb in main
	    #13 0x79a1f85d9082 in __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/libc-start.c:308:16
	    #14 0x589ab43a93bd in _start
```

**Pull request merged**
